### PR TITLE
Testing chapter

### PR DESCRIPTION
**Pull request by bebraw**
Do you think the techniques should go to another chapter? Now there's a high level introduction and I can expand on that a bit. Doing this would give us two 1-2k word chapters on testing - one theoretical, one practical.

### CODE REVIEW

File: manuscript/code-quality/02-testing.md [hunk 1]
```diff
@@ -1,14 +1,42 @@
 # Testing
 
-TODO
+When you are developing code, you will **test** it either by inspecting the result manually or by writing a script that **asserts** the truths you expect. If you rely on manual inspection, maintaining code becomes harder the more code you write as you have more truths to inspect and you could miss a check during this process.
```
Comment (sapegin) [inline manuscript/code-quality/02-testing.md]: I’d add two testing pyramids here: traditional (unit > integration > UI) and modern (linters > types > tests).

File: manuscript/code-quality/02-testing.md [hunk 2]
```diff
@@ -1,14 +1,42 @@
...
+
+## The Value of Typing
```
Comment (sapegin) [inline manuscript/code-quality/02-testing.md]: I think this section is too abstract, especially the second paragraph. How can use this information as a JavaScript developer?

File: manuscript/code-quality/02-testing.md [hunk 3]
```diff
@@ -1,14 +1,42 @@
...
+
+Especially when you are prototyping an idea, having a loose type system can be beneficial as it doesn't get into your way. As you understand what you are doing better, ideally you can model stricter typing to fit the problem and to solidify the code against type related errors.
+
+Type errors are only one possible problem, though. Your code can compile while still do the wrong thing. The algorithm could be incorrect or the code performs operations that mutate the application state in such way that it breaks and fails with a runtime error. Certain environments, such as Elm, disallow runtime errors altogether by design.
+
+## Testing is Specification
+
+Testing is about **specifying** what you want. The work of specifying can be performed on different levels of abstraction. If you look at the code from developer perspective, you can think about testing in terms of **unit testing**, **property based testing**, or **integration testing**. If you test from the user perspective, you get into the realm of **acceptance testing**. Regression, performance, and mutation testing extend the scope further.
```
Comment (sapegin) [inline manuscript/code-quality/02-testing.md]: And this is even more abstract. I’d remove the first paragraph (all it says: there are many types of tests and all have pros and cons) and expand the second — explain the difference between different types of tests.

File: manuscript/code-quality/02-testing.md [hunk 4]
```diff
@@ -1,14 +1,42 @@
...
+The techniques look at the same problem of specification from different angles that overlap sometimes. They also have varying degrees of difficulty. Whereas a unit test is cheap to write, figuring out a good property test against can be harder but also more powerful. Acceptance testing against a user interface comes with additional complexity as you have to mimic the user behavior accurately.
+
+## How Much to Test?
+
+Regardless of how much you test, manual testing may still be required especially if you target multiple platforms. But it's a good goal to minimize the amount of it and even try to eliminate it.
 
-* Why
-* What's the point/value
-* What to test
+No matter how much you test, problems may still slip through. For this reason, it's good to remember that you get what you measure. If you want performance, you should test against performance. If you want to avoid regressions against downstream projects, you should test against them before publishing changes.
```
Comment (sapegin) [inline manuscript/code-quality/02-testing.md]: What are *downstream projects*? Projects that depend on yours?

File: manuscript/code-quality/02-testing.md [hunk 5]
```diff
@@ -1,14 +1,42 @@
...
 
-TODO: Run tests of popular users of your library with a new version to test that it doesn’t break them
+Tests come with a cost. Test code is code you have to maintain in addition to the implementation. Having more tests isn't necessarily better. It's good to put conscious effort into maintaining good tests. Fortunately tests allow refactoring and sometimes finding better ways to implement the desired features can simplify both the implementation and tests.
```
Comment (sapegin) [inline manuscript/code-quality/02-testing.md]: Tests can be very slow as well, so slow that developers won’t run them. Especially UI tests.

Tests can be super hard to maintain (for example those terrible super long XPath selectors in UI tests), so developers will comment them out instead of fixing.

File: manuscript/code-quality/02-testing.md [hunk 6]
```diff
@@ -17,9 +45,8 @@ TODO
 * Show how to implement unit tests through `assert` -> how to evolve a testing framework of your own
```
Comment (sapegin) [inline manuscript/code-quality/02-testing.md]: “how to evolve a testing framework of your own” — feel out of scope of the book on how to simplify maintenance ;-)

File: manuscript/code-quality/02-testing.md (hunk 4 above)
Comment (bebraw) [inline manuscript/code-quality/02-testing.md]: Correct. I think we have to define *upstream* and *downstream* somewhere as package exist as a part of ecosystem somehow and that has implications.

**Comment by sapegin** [comment]
> Do you think the techniques should go to another chapter?

I think it’s a good idea, I’d like to see the first part more independent. Now I have more questions then answers after reading it ;-)

File: manuscript/code-quality/02-testing.md (hunk 6 above)
Comment (bebraw) [inline manuscript/code-quality/02-testing.md]: Yup, probably better to drop that idea. Too far off.

File: manuscript/code-quality/02-testing.md (hunk 1 above)
Comment (bebraw) [inline manuscript/code-quality/02-testing.md]: Do you have some reference for the latter? I have to understand the parallel better.

Comment (sapegin) [inline manuscript/code-quality/02-testing.md]: It’s here: https://www.youtube.com/watch?v=m_vUUgI0bo8&t=24361s

Linters and types are actually at the same level ;-|

Comment (bebraw) [inline manuscript/code-quality/02-testing.md]: Ok, I think what will happen is that I'll split this up. I'll push the proposed split in a bit.

**Comment by bebraw** [comment]
The testing chapter could use another look. I ended up explaining the techniques in the chapter itself with a proper example here and there to get to the point.

It's probably better to work on the other chapters in separate PRs since this PR improves the structure of the book.

File: manuscript/code-quality/04-testing.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+Regression testing applies on API level as well. An API that has worked in the past should work the same unless the API contract changed and a new major version was released. Even then it may be a good idea to provide a deprecation path for users so they can update their software to the new API.
+
+T> [dont-break](https://www.npmjs.com/package/dont-break) is a good example of a package meant for testing your project against downstream (dependent) projects.
+
+### Performance Testing
+
+Given you get what you measure, **performance testing** is valuable for projects that care about speed. Packages like [k6](https://k6.io/), [Benchmark.js](https://benchmarkjs.com/) and [matcha](https://www.npmjs.com/package/matcha) can be helpful here. You can benchmark your implementation against competing libraries or an older version of yours.
+
+Ideally, you should store the old results somewhere and keep the runtime environment the same across separate runs. If the environment varies between runs, benchmarking results aren't comparable.
```
Comment (sapegin) [inline manuscript/code-quality/04-testing.md]: I think it’s not possible with conventional CI servers. So better to run tests for old and new code and compare the difference.

File: manuscript/code-quality/04-testing.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+To make sure your npm package paths (`main`, `module`, i.e.) work, you can validate your package through [pkg-ok](https://www.npmjs.com/package/pkg-ok).
+
+?> Cover other package testing tools here.
+
+### Code Coverage
+
+**Code coverage** tells you how much of the code is covered by tests. The technique can reveal parts of the code that aren't being tested. Those parts have unspecified behavior and as a result may contain issues. Code coverage doesn't tell anything about the quality of the tests, but it can be used to discover places to test.
+
+[Codecov](https://codecov.io/) and [Coveralls](https://coveralls.io/) provide code coverage and a badge to include in your README. They also collect coverage for every pull request and post a comment explaining how a pull request would affect project coverage as you should maintain the current coverage at a minimum and ideally improve it.
```
Comment (sapegin) [inline manuscript/code-quality/04-testing.md]: They don’t really provide coverage themselves, they publish reports provided by your coverage framework ;-)

So worth adding Istanbul (and Jest that includes it by default).

File: manuscript/code-quality/03-typing.md [hunk 9]
```diff
@@ -0,0 +1,25 @@
+# Typing
+
+A function interface is a contract. Depending on the system, it gives you guarantees. In JavaScript there are no guarantees by default and you can pass anything to a function. This may lead to a runtime error, though, given it's an interpreted language. In a stricter system passing invalid values would be forbidden on a type level.
+
+## The Value of Typing
```
Comment (bebraw) [inline manuscript/code-quality/03-typing.md]: I reworked it a bit. We have to look into the typing chapter carefully in a separate PR. Especially the Flow and TypeScript bits are challenging. We have to show how to get started and how they differ.

File: manuscript/code-quality/04-testing.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+
+### Unit Testing
+
+**Unit testing** asserts specific truths about code. Each unit should verify the single truth. You can get started by using Node [assert](https://nodejs.org/api/assert.html). Specialized assertion libraries have more functionality and better output but `assert` is a good starting point.
+
+The first step is to include a test within the code:
+
+**add.js**
+
+<!-- textlint-disable -->
```
Comment (bebraw) [inline manuscript/code-quality/04-testing.md]: I think I have to tweak the linting rules. It's choking on `const`.

File: manuscript/code-quality/04-testing.md [hunk 11]
```diff
@@ -0,0 +1,205 @@
+# Testing
+
+Tests can be seen as a runnable specification that describes truths about the implementation. Automated testing is a form of risk management and reduces risks related to the project. Manual testing is the other end of the spectrum. It's also the most labor intensive and brittle option.
+
+On an operational level, **analytics** can be seen as testing. If you capture information about how the project is being used, you can use this information to improve the project and focus on the right things. Doing this allows you to build the right things instead of building them right.
```
Comment (bebraw) [inline manuscript/code-quality/04-testing.md]: Yeah, good point. I think I'll add some examples as bullets so you get a better idea of what value testing might provide. Not only technical, that's true for sure.

File: manuscript/code-quality/04-testing.md [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+const add = require('./add');
+
+assert(add(1, 1) === 2);
+```
+
+<!-- textlint-enable -->
+
+If you run `node ./add.test.js`, you should have the same situation as before. A good test runner would pick up the wanted tests and run them like this. [globbing](https://www.npmjs.com/package/glob) would be one way.
+
+Fortunately, people have written a lot of good test runners and frameworks, so you don't have to do this yourself. Particularly [Jest](https://www.npmjs.com/package/jest), [Mocha](https://www.npmjs.com/package/mocha), and [tape](https://www.npmjs.com/package/tape) are good starting points.
```
Comment (bebraw) [inline manuscript/code-quality/04-testing.md]: Yeah, I'll refactor. 👍

File: manuscript/code-quality/04-testing.md (hunk 10 above)
Comment (sapegin) [inline manuscript/code-quality/04-testing.md]: What it says? It shouldn’t lint anything, only format code.

Comment (bebraw) [inline manuscript/code-quality/04-testing.md]: `34:1  error  Parsing error: The keyword 'const' is reserved  eslint`

I'll try tuning the configuration.

Comment (bebraw) [inline manuscript/code-quality/04-testing.md]:
```
  "env": {
    "es6": true
  },
```

is enough. I'll go with that.

Comment (sapegin) [inline manuscript/code-quality/04-testing.md]: 👍

**Comment by bebraw** [comment]
I added bits as discussed. Thanks for the feedback. 👍 

This should be good for another review. We can continue on those missing bits and TODOs in separate PRs after this lands.

File: manuscript/code-quality/03-typing.md [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+Loose typing comes with the following problems:
+
+* It allows runtime errors to exist. If you pass a value of wrong type, it can lead to unexpected results and crash the entire system.
+* Even if you knew the expected type, communicating it would be difficult. You can perform a type check on language level but the problem with this is that it's difficult to use the type information in tooling then.
+* Types can be annotated within comments but then you would lose code level checks without compiling the comments to actual checks.
+
+Certain languages allow **type hints** as a compromise. If you know a type, add it to the code. The language uses this information when available. One side benefit of stating types this way is that then the interpreter can use these guarantees to speed up execution. The type information can be used for **property based testing** and documentation.
+
+Flow type checker is an example of a solution that builds on top of JavaScript while providing stronger typing semantics for it. TypeScript is an entire language that comes with a typing approach of its own.
```
Comment (sapegin) [inline manuscript/code-quality/03-typing.md]: TS isn’t really a language. The language is still ECMAScript. The only difference as I see it is that TS has their own compiler.
